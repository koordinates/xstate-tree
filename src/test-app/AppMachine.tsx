import { identity } from "lodash";
import React from "react";
import { createMachine } from "xstate";

import {
  buildView,
  buildXStateTreeMachine,
  buildRootComponent,
  singleSlot,
  buildActions,
  lazy,
} from "../";
import { Link, RoutingEvent } from "../routing";

import { TodosMachine } from "./TodosMachine";
import { homeRoute, settingsRoute, history } from "./routes";

// eslint-disable-next-line @typescript-eslint/ban-types
type Context = {};
type Events =
  | RoutingEvent<typeof homeRoute>
  | RoutingEvent<typeof settingsRoute>;
type State =
  | { value: "todos"; context: Context }
  | { value: "otherScreen"; context: Context };
const ScreenSlot = singleSlot("ScreenGoesHere");
const slots = [ScreenSlot];
const OtherMachine = () =>
  import("./OtherMachine").then(({ OtherMachine }) => OtherMachine);
const AppMachine = createMachine<Context, Events, State>(
  {
    id: "app",
    initial: "waitingForRoute",
    on: {
      GO_HOME: {
        target: ".todos",
        cond: (_ctx, e) => e.meta.indexEvent ?? false,
      },
      GO_SETTINGS: ".otherScreen",
    },
    states: {
      waitingForRoute: {},
      todos: {
        invoke: {
          id: ScreenSlot.getId(),
          src: "TodosMachine",
        },
      },
      otherScreen: {
        invoke: {
          id: ScreenSlot.getId(),
          src: "OtherMachine",
        },
      },
    },
  },
  {
    services: {
      TodosMachine: TodosMachine,
      OtherMachine: lazy(OtherMachine),
    },
  }
);

const actions = buildActions(AppMachine, identity, () => ({}));

const AppView = buildView(
  AppMachine,
  identity,
  actions,
  slots,
  ({ slots, inState }) => {
    return (
      <>
        {inState("todos") && (
          <>
            <p data-testid="header">On home</p>
            <Link to={settingsRoute} testId="swap-to-other-machine">
              Swap to settings
            </Link>
          </>
        )}
        {inState("otherScreen") && (
          <>
            <p data-testid="header">On settings</p>
            <Link to={homeRoute} testId="swap-to-other-machine">
              Swap to home
            </Link>
          </>
        )}
        <slots.ScreenGoesHere />
      </>
    );
  }
);

export const BuiltAppMachine = buildXStateTreeMachine(AppMachine, {
  actions,
  selectors: identity,
  slots,
  view: AppView,
});

export const App = buildRootComponent(BuiltAppMachine, {
  history,
  basePath: "",
  routes: [homeRoute, settingsRoute],
  getPathName: () => "/",
  getQueryString: () => "",
});
