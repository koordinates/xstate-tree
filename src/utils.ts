import { ComponentPropsWithRef, JSXElementConstructor } from "react";
import { Interpreter, StateMachine } from "xstate";

export type PropsOf<
  C extends keyof JSX.IntrinsicElements | JSXElementConstructor<any>
> = JSX.LibraryManagedAttributes<C, ComponentPropsWithRef<C>>;

export function delay(ms = 0): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export type OmitOptional<T> = {
  [P in keyof Required<T> as Pick<T, P> extends Required<Pick<T, P>>
    ? P
    : never]: T[P];
};
export type IsEmptyObject<Obj, ExcludeOptional extends boolean = false> = [
  keyof (ExcludeOptional extends true ? OmitOptional<Obj> : Obj)
] extends [never]
  ? true
  : false;

export function assertIsDefined<T>(
  val: T,
  msg?: string
): asserts val is NonNullable<T> {
  if (val === undefined || val === null) {
    throw new Error(
      `Expected 'val' to be defined, but received ${val} ${
        msg ? `(${msg})` : ""
      }`
    );
  }
}

export function assert(value: unknown, msg?: string): asserts value {
  if (typeof expect !== "undefined") {
    if (value !== true && msg) {
      console.error(msg);
    }
    expect(value).toEqual(true);
  } else if (value !== true) {
    if (msg) {
      console.error(msg);
    }
    throw new Error("assertion failed");
  }
}

export type StateMachineToInterpreter<T> = T extends StateMachine<
  infer TContext,
  infer TSchema,
  infer TEvents,
  infer TState,
  any,
  any,
  any
>
  ? Interpreter<TContext, TSchema, TEvents, TState, any>
  : never;
