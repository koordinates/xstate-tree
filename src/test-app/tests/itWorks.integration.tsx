import { render } from "@testing-library/react";
import React from "react";

import { delay } from "../../utils";
import { App } from "../AppMachine";

describe("Test app", () => {
  it("renders the initial app", async () => {
    const { container } = render(<App />);

    await delay(100);
    expect(container).toMatchSnapshot();
  });
});
