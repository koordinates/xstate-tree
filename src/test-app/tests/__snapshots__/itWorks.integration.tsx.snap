// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test app renders the initial app 1`] = `
<div>
  <p
    data-testid="header"
  >
    On home
  </p>
  <a
    data-testid="swap-to-other-machine"
    href="/settings/"
  >
    Swap to settings
  </a>
  <header
    class="header"
  >
    <h1>
      todos
    </h1>
    <input
      class="new-todo"
      data-testid="todo-input"
      placeholder="What needs to be done?"
      value=""
    />
  </header>
  <section
    class="main"
  >
    <input
      class="toggle-all"
      data-testid="update-all"
      id="toggle-all"
      type="checkbox"
    />
    <label
      for="toggle-all"
    >
      Mark all as complete
    </label>
    <ul
      class="todo-list"
    >
      <li
        class=""
        data-testid="todo"
      >
        <div
          class="view"
        >
          <input
            class="toggle"
            data-testid="toggle-todo"
            type="checkbox"
          />
          <label>
            foo
          </label>
          <button
            class="destroy"
            data-testid="remove-todo"
          />
        </div>
        <input
          class="edit"
          value="foo"
        />
      </li>
      <li
        class="completed"
        data-testid="todo"
      >
        <div
          class="view"
        >
          <input
            checked=""
            class="toggle"
            data-testid="toggle-todo"
            type="checkbox"
          />
          <label>
            bar
          </label>
          <button
            class="destroy"
            data-testid="remove-todo"
          />
        </div>
        <input
          class="edit"
          value="bar"
        />
      </li>
    </ul>
  </section>
  <footer
    class="footer"
  >
    <span
      class="todo-count"
    >
      <strong
        data-testid="uncompleted-count"
      >
        1
      </strong>
       
      item
       left
    </span>
    <ul
      class="filters"
    >
      <li>
        <a
          class="selected"
          data-testid="view-all"
          href="#/"
        >
          All
        </a>
      </li>
      <li>
        <a
          class=""
          data-testid="view-active"
          href="#/active"
        >
          Active
        </a>
      </li>
      <li>
        <a
          class=""
          href="#/completed"
        >
          Completed
        </a>
      </li>
    </ul>
    <button
      class="clear-completed"
    >
      Clear completed
    </button>
  </footer>
</div>
`;
